T2D-215 — Make all objects associated to a study public only after a given date

tskir opened the pull request:
### [Diff link](https://github.com/EBIvariation/metadata-ws/pull/53/files)

This implements release date control for all Auditable entities:

* Study — `releaseDate` field set directly;
* Analysis, WebResource — 1..M from Study, release date is calculated by fetching the field from the parent study;
* File, Publication, ReferenceSequence, Sample, Taxonomy — M..M, release date is calculated via a `@Formula` inner join query.

Fixed how LocalDate entities are stored in the SQL by adding `hibernate-java8` dependency.

The Cucumber tests are included and passing as well. They include all combinations of unset/past/future release dates for study, its direct and indirect child entities.

Also added documentation for how the transitive release date control works.

tskir commented:
To avoid unnecessary rebases and to make reading diffs easier, the target branch of this PR is https://github.com/EBIvariation/metadata-ws/pull/50. I will change is to `master` just before we do the actual merge.

tskir commented:
To avoid unnecessary rebases and to make reading diffs easier, the target branch of this PR is https://github.com/EBIvariation/metadata-ws/pull/50. I will change is to `master` just before we do the actual merge.

iyangar commented:
Please fix build fail

tskir commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/File.java:
```diff
@@ -123,4 +125,16 @@ public long getSize() {
         return type;
     }
+
+    // Release date control: get the *earliest* release date from all studies which link to this file
```
As @iyangar noted during one of the standups, the nature of transitive release date control should be documented. Will do this as well.

tskir replied:
Documentation added.

cyenyxe replied:
Could you please convert it to Javadoc? It basically describes what the field is. Applies to other classes.

tskir replied:
Good suggestion, fixed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/85037d329878a7903955a76d5efd5842287cf87b.

tskir commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/aop/StudyReleaseDateAspect.java:
```diff
@@ -42,39 +41,44 @@
+    @Around("execution(* uk.ac.ebi.ampt2d.metadata.persistence.repositories.*.findAll(..))")
+    public Object filterOnReleaseDateAdviceFindAll(ProceedingJoinPoint proceedingJoinPoint) throws Throwable {
+        Iterable<Object> results = (Iterable<Object>) proceedingJoinPoint.proceed();
+        Iterator<Object> i = results.iterator();
```
Yes, because the `findAll` methods of all repositories return:
* Either **`Iterable`**,
* Or `Page`, which extends `Slice`, which extends **`Iterable`**.

So we're always safe to use the `Iterable` methods.

tskir commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/aop/StudyReleaseDateAspect.java:
```diff
@@ -17,14 +17,13 @@
+import java.util.Iterator;
 
 /**
  * An @Aspect for ensuring only published studies are returned through checking a study's releaseDate field
```
Fixed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/0934a756cf59798fe498d1d26c16389a1c498009. Updated the comment, renamed the class (to `ReleaseDateAspect`) as well as all associated methods and documentation strings.

tskir commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/aop/StudyReleaseDateAspect.java:
```diff
@@ -42,39 +41,44 @@
+    private Object processReleaseControlledObject(Object result) {
+        if (result instanceof Auditable) {
+            LocalDate releaseDate = ((Auditable) result).getReleaseDate();
+            if (releaseDate != null && releaseDate.isAfter(LocalDate.now())) {
```
I personally think that `null` is always available would be more reasonable choice, but of course it's for @cyenyxe to decide.

Blank release date is now included in the test cases, fixed.

iyangar replied:
sure, we can have some specific value for infinite if necessary.

tskir commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/aop/StudyReleaseDateAspect.java:
```diff
@@ -42,39 +41,44 @@
-     * object is not a Study object, return as it is. If the returned object is a Study object, check the study's
-     * releaseDate field. Return null if the releaseDate is a date in the future.
+     * object is not a Auditable object, return as it is. If the returned object is an Auditable object, check its
+     * getReleaseDate() method. Return null if the release date is null (i. e., not set) or a date in the future.
```
Fixed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/0934a756cf59798fe498d1d26c16389a1c498009, now the documentation only refers to the “release date”

tskir commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Auditable.java:
```diff
@@ -42,4 +43,6 @@ public void setLastModifiedDate(ZonedDateTime lastModifiedDate) {
 
     public abstract ID getId();
+
+    public LocalDate getReleaseDate() {return null;}
```
In https://github.com/EBIvariation/metadata-ws/pull/53/commits/0934a756cf59798fe498d1d26c16389a1c498009 fixed formatting to next line, and also all places where a space was missing

tskir commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/WebResource.java:
```diff
@@ -21,18 +21,10 @@
-import javax.persistence.SequenceGenerator;
-import javax.persistence.Table;
-import javax.persistence.UniqueConstraint;
+import javax.persistence.*;
```
Thank you, I only recently noticed that IDEA is replacing imports with asterisks, and it's really annoying. I have now disabled it and replaced all occurrences. Fixed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/0934a756cf59798fe498d1d26c16389a1c498009.

selvaebi replied:
this class is not fixed yet.

tskir replied:
@selvaebi You're right! I fixed everything but this one. Now finally fixed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/a46b2eba672b0fed1e11bf497f2ff023398b57fe. I double checked that no more instances of this import pattern remain.

tskir commented on metadata-ws/src/test/java/uk/ac/ebi/ampt2d/metadata/cucumber/StudySteps.java:
```diff
@@ -41,6 +41,13 @@
     @Autowired
     private MockMvc mockMvc;
 
+    @When("I create a study with (.*) for taxonomy and (.*) for accession$")
```
This is a very good point since it reduces code duplication. Now fixed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/0934a756cf59798fe498d1d26c16389a1c498009.

tskir commented on metadata-ws/src/test/java/uk/ac/ebi/ampt2d/metadata/cucumber/StudySteps.java:
```diff
@@ -41,6 +41,13 @@
     private MockMvc mockMvc;
 
+    @When("I create a study with (.*) for taxonomy and (.*) for accession$")
+    public void createTestStudyAccession(String testTaxonomyKey, String accession) throws Exception {
```
Renamed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/0934a756cf59798fe498d1d26c16389a1c498009.

tskir commented on metadata-ws/src/test/resources/features/analysis.feature:
```diff
@@ -398,3 +398,40 @@ Feature: analysis object
+    When I request GET with value of ANALYSIS
+    Then the response code should be 200
+
+    Given I request PATCH STUDY with patch and day tomorrow
```
Could you please elaborate? To clarify, this is not the point where we check availability, this is where we **modify** the `releaseDate` of a Study. And PATCH operation is definitely being done on the resource, see: https://github.com/tskir/metadata-ws/blob/d1ed96e2495be12fe5dee5a1008737e0689db0c5/metadata-ws/src/test/java/uk/ac/ebi/ampt2d/metadata/cucumber/CommonSteps.java#L182

iyangar replied:
ok, i didn't observe

tskir commented on metadata-ws/src/test/resources/features/file.feature:
```diff
@@ -34,3 +34,74 @@ Feature: file object
+    When I request PATCH ANALYSIS2 with list FILE of files
+    Then the response code should be 2xx
+
+    Given I request PATCH STUDY1 with patch and day tomorrow
```
Similarly, see my reply above

tskir commented on metadata-ws/src/test/resources/features/file.feature:
```diff
@@ -34,3 +34,74 @@ Feature: file object
     When I request GET /files
     Then the response should contain one file
+
+  Scenario: attempt to get a file of an analysis of an unreleased study must fail
```
The tests have now been significantly updated and expanded. They now include all combinations of null/past/future release dates for child entities of a Study.

The Study itself does not need to be tested because there are already existing tests in `study.feature` which check everything about Study's `releaseDate`

tskir commented:
> Please fix build fail

Thank you @iyangar for a thorough review. This is now fixed as well, the build is passing.

tskir commented:
Implementation for all `Auditable` entities is now ready. Added docs, significantly amended and expanded the tests, fixed persisting `LocalDate` objects to the database (it was stored as a blob before, not suitable for querying), improved code style, formatting and variable/class naming. Also the internal `releaseDate` field is now hidden from API responses except for the `Study` entity.

Ready for the next iteration of review.

selvaebi commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Taxonomy.java:
```diff
@@ -75,4 +78,19 @@ public Long getId() {
+            "INNER JOIN study on analysis.study_id = study.id " +
+            "WHERE sample.id=id)")
+    @JsonIgnore
+    private LocalDate releaseDate;
```
I guess the taxonomy and reference sequence will be public. not sure can check with @cyenyxe

cyenyxe replied:
Yes, none of our archives supports non-public taxonomies and sequences.

tskir replied:
Fixed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/4b70202f0d05260d4268461280280e1d5296aa10. Taxonomy and WebResource are now always public.

selvaebi commented on metadata-ws/src/test/resources/features/release-date.feature:
```diff
@@ -0,0 +1,124 @@
+Feature: Transitive release date control for child objects of a Study
```
This looks good please add tests for publications and web resources as well. and also patching the unreleased study to released and vice-versa.

tskir replied:
Fixed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/f2d050aede91f1fa82253e24924f6e68a86fa038. Expanded the tests to:
- Check for Publication and WebResource
- Test patching the study from released to unreleased state

Unfortunately we can't test patching unreleased to released, because, as I mentioned during one of the previous standups, the moment the study we make study unreleased it becomes inaccessible (because it is, well, unreleased). This will be fixed when T2D-216 (Allow admins and object owners to access them before release date) is implemented.

selvaebi replied:
please see my comments in ReleaseDateAspect file

tskir replied:
Added tests for patching Study from unreleased to released and vice versa in https://github.com/EBIvariation/metadata-ws/pull/53/commits/8ffa5dcca48166a32ac23b29ad55df341f95f290.

cyenyxe commented on metadata-ws/src/test/resources/features/release-date.feature:
```diff
@@ -0,0 +1,124 @@
+    Then set the URL to TAXONOMY
+    And the response code should be 201
+
+    # Create the common reference sequences
```
This block should be like:

1. When I request POST /reference-sequences with JSON-like payload (...)
2. And set the URL (...)
3. Then the response code should be 201

This template can be used as reference for others below.

tskir replied:
Thank you for the clarification, looking at existing tests I couldn't quite determine what is the accepted practice. Now rewrote, fixed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/2da2671086076bae4c89bb2b6f79b4363089ef75.

cyenyxe commented on metadata-ws/src/test/resources/features/release-date.feature:
```diff
@@ -0,0 +1,124 @@
+    Then set the URL to REFERENCE_SEQUENCE
+    And the response code should be 201
+
+    # Create two studies with the release date of "yesterday"
```
This block should be the following succession:

1. When
2. And
3. Then
4. When
5. And
6. Then

This template can be used as reference for others below.

tskir replied:
Fixed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/2da2671086076bae4c89bb2b6f79b4363089ef75.

cyenyxe commented on metadata-ws/src/test/resources/features/release-date.feature:
```diff
@@ -0,0 +1,124 @@
+    When I request PATCH ANALYSIS2 with list FILE2,FILE3 of files
+    Then the response code should be 2xx
+
+    # Set release dates for studies
```
This block should be:

1. When
2. Then
3. When
4. Then

tskir replied:
Fixed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/2da2671086076bae4c89bb2b6f79b4363089ef75.

tskir commented on docs/release-date.md:
```diff
@@ -0,0 +1,18 @@
+* For all other **Auditable** entities:
+  + If at least one released Study links to an entity (including transitive linking, e. g. Study linking to Analysis
+    linking to File), this entity is considered to be released and becomes retrievable through API.
+  + Otherwise, the entity is not released and not retrievable through API.
```
Thank you, fixed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/79bae834505866afe20d6efd2582a2ea9369a11c

tskir commented on docs/release-date.md:
```diff
@@ -0,0 +1,18 @@
+    child studies in either direction.
+* For all other **Auditable** entities:
+  + If at least one released Study links to an entity (including transitive linking, e. g. Study linking to Analysis
+    linking to File), this entity is considered to be released and becomes retrievable through API.
```
Fixed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/79bae834505866afe20d6efd2582a2ea9369a11c

tskir commented:
Rebased the PR on top of recent changes in #47 and #50. Implemented Study M..M WebResource relation, restructured and expanded the tests, modified documentation and comments. Ready for the third iteration of review.

selvaebi commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/aop/ReleaseDateAspect.java:
```diff
@@ -17,64 +17,67 @@
      * @throws Throwable
      */
-    @Around("execution(* org.springframework.data.repository.CrudRepository.findOne(java.io.Serializable))")
+    @Around("execution(* uk.ac.ebi.ampt2d.metadata.persistence.repositories.*.findOne(java.io.Serializable))")
```
It should be crudRepository only, if not we can't patch an unreleased study.

tskir replied:
You're right, thank you! Fixed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/8ffa5dcca48166a32ac23b29ad55df341f95f290.

selvaebi commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Study.java:
```diff
@@ -209,11 +205,31 @@ public void setResources(List<WebResource> resources) {
 
+    public void setResource(WebResource resource) {
+        List<WebResource> resources = this.getResources();
+        if (resources != null) {
```
```suggestion - can be changed to 
        if (resources == null) {
        resources = new ArrayList<>();
        }
        resources.add(resource);
```

tskir replied:
Thanks, fixed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/f50f715fb67cb7453c0ca920d5ddfecf3a3a37da.

selvaebi commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Taxonomy.java:
```diff
@@ -17,8 +17,10 @@
+import com.fasterxml.jackson.annotation.JsonIgnore;
 import com.fasterxml.jackson.annotation.JsonProperty;
 import io.swagger.annotations.ApiModelProperty;
+import org.hibernate.annotations.Formula;
```
can be removed.

tskir replied:
You're right, removed this and other unused annotations in https://github.com/EBIvariation/metadata-ws/pull/53/commits/f50f715fb67cb7453c0ca920d5ddfecf3a3a37da.

selvaebi commented on metadata-ws/src/test/resources/features/release-date.feature:
```diff
@@ -0,0 +1,174 @@
+      # File is available WHEN AND ONLY WHEN at least one study which links to it is available
+      # Release dates of "null" or "yesterday" should result in a RELEASED state;
+      # "tomorrow" in an UNRELEASED state.
+      | S1_RELEASE | S2_RELEASE | A1  | A2  | F1  | F2  | F3  | P1  | W2  |
```
sorry can this be done for today as well ?

tskir replied:
No problem at all, added `today` to list of checked values. Now every combination of `none`, `yesterday`, `today` and `tomorrow` is tested. Fixed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/588c05d0d135683d056d38b461f0194e486c185d.

tskir commented:
Rebased this PR on top of recent changes in #47 and #50. Addressed review comments:
* Code style changes
* Added test for patching released Study to unreleased and vice versa
* Added `today` to list of checked values in release date tests

**Now ready for the fourth iteration of review.**

tskir commented:
Rebased on top of recent changes in #47 and #50. Made a minor change to `Study.java` (there was an `setResource()` method which wasn't used anywhere and could also lead to the introduction of duplicate entries in the list due to flawed implementation).

**Ready for the fifth iteration of review.** [_Diff link_](https://github.com/EBIvariation/metadata-ws/pull/53/files)

tskir commented:
@cyenyxe 
Rebased on top of recent changes in #50, ready for review

tskir commented:
@cyenyxe Rebased on top of the recent changes in #50, ready for review

cyenyxe commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/aop/ReleaseDateAspect.java:
```diff
@@ -69,12 +68,16 @@ public Object filterOnReleaseDateAdviceFindAll(ProceedingJoinPoint proceedingJoi
-        if ( result == null ||
-                (result instanceof Study && ((Study) result).getReleaseDate().isAfter(LocalDate.now())) ) {
-            return null;
+    private Object processReleaseControlledObject(Object result) {
```
I would suggest using a name that states more clearly the functionality of the method, such as `getObjectUnlessBeforeReleaseDate`.

tskir replied:
Agreed, fixed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/b12ad94aa306c0e8eb84f967c15b5cb6baa5fc2c

cyenyxe commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/Auditable.java:
```diff
@@ -42,4 +43,8 @@ public void setLastModifiedDate(ZonedDateTime lastModifiedDate) {
 
     public abstract ID getId();
+
+    public LocalDate getReleaseDate() {
```
Wouldn't this be best as a pure virtual method, so all the `Auditable` classes are forced to implement it? The ones like `ReferenceSequence`, whose release date is not explicitly needed, would have to implement it just to return null; but on the other hand we wouldn't have bogus default implementations that return null when they shouldn't.

tskir replied:
I agree that this is better and more explicit in case of “always public” entities. Implemented `getReleaseDate()` as abstract method, changed `Taxonomy` and `ReferenceSequence` classes accordingly.

cyenyxe commented on metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/ReferenceSequence.java:
```diff
@@ -17,8 +17,10 @@
  */
 package uk.ac.ebi.ampt2d.metadata.persistence.entities;
 
+import com.fasterxml.jackson.annotation.JsonIgnore;
```
The newly added imports don't seem to be necessary in this class.

tskir replied:
Removed all unnecessary imports in this and other entity classes, fixed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/b12ad94aa306c0e8eb84f967c15b5cb6baa5fc2c

cyenyxe commented on metadata-ws/src/test/resources/features/release-date.feature:
```diff
@@ -0,0 +1,191 @@
+    }
+    """
+    Then the response code should be 201
+    And set the URL to FILE3
```
There is no `then` after this `and`, so no additional check is performed. Is that intentional?

tskir replied:
Not intentional, `Then` and `And` were accidentally swapped around. Now fixed in https://github.com/EBIvariation/metadata-ws/pull/53/commits/b12ad94aa306c0e8eb84f967c15b5cb6baa5fc2c
